ref_arch/multi_tenant.rst

```diff
@@ -0,0 +1,393 @@
+
+The order/limit query slightly favors distribution by :code:`company_id`. However JOIN queries differ more dramatically.
+
+.. note::
+
+  In our normalized schema above, the ads table does not have a company_id column because it can retrieve that information through the campaigns table. If we want to distribute the ads table by company id however, we would need to denormalize the schema slightly and add that column. The query below assumes we have done this, and we'll talk more about this technique later.
+
+::
+
+  -- running campaigns which receive the most clicks and impressions
+  -- for a single tenant
+
+  SELECT campaigns.id, campaigns.name, campaigns.monthly_budget,
+         sum(impressions_count) as total_impressions,
+         sum(clicks_count) as total_clicks
+  FROM ads, campaigns
+  WHERE ads.company_id = campaigns.company_id
+  AND campaigns.company_id = 5
+  AND campaigns.state = 'running'
+  GROUP BY campaigns.id, campaigns.name, campaigns.monthly_budget
+  ORDER BY total_impressions, total_clicks;
+
+DIAGRAM: show id repartitioning, and company_id routing
+
+For this query, distributing by campaign id is quite bad. Workers must use a lot of network traffic to pull related information together for the join, in a process called *repartitioning.* Routing the query for execution in a single worker avoids the overhead, and is possible when distributing by :code:`company_id`. The placement of related information together on a worker is called *co-location.*
+
+These queries indicate a general design pattern: distributing shards by tenant id (such as the company id) allows Citus to route queries to individual workers for efficient processing. This fits multi-tenant applications which join structured information together per-tenant.
+
+Preparing Tables for Distribution
+---------------------------------
+
+In the previous section we identified the correct distribution column for multi-tenant applications: the tenant id. We also saw that some tables designed for a single machine PostgreSQL instance may need to be denormalized by the addition of this column.
+
+We will need to modify our schema, but there is one other caveat to note about distributed systems. Enforcing uniqueness or foreign key constraints in Citus requires that they include the distribution column. Our tables don't currently do that, for instance in the ads table we specify
+
+::
+
+  -- not efficiently enforceable
+
+  campaign_id bigint REFERENCES campaigns (id)
+
+This constraint includes only the campaign id, not the company (tenant) id. In order to verify the constraint Citus might have to consult multiple workers because it's not guaranteed in all situations that the ad in question is co-located with its campaign.
+
+To guarantee that they are co-located, ad and campaign must both be distributed by company_id, and this column must appear in the foreign key. Similarly the primary key, implying uniqueness as it does, must be composite and include company_id.
```

Comment:
I think we should break this into pieces:-

(a) Explain why we need to have company id on all tables. This could follow from the description I have above or some modification of it.

(b) Explain changes to primary / foreign keys assuming schema changes in (a). i.e. If we don't include the tenant id in the primary key, Citus will need to check all worker nodes to enforce uniqueness on each index, which does not scale well. Adding tenant id to your primary keys allows Citus to ensure this at a shard level which can be completely enforced on the worker nodes.